### PR TITLE
Refactor the particle-particle broad search code in DEM

### PR DESCRIPTION
**Pull request by acdaigneault**
# Description of the problem

- PR n°1 of a series of refactoring PRs for the DEM code.
- Enhancement of the particle-particle broad search code.

# Description of the solution

- Addition of the private method `particle-particle broad search code` that allows to fill containers of contact pair candidates with a range of particles and the first iterator of this range (prior skipping the first particle when used for particles in main cell).

### CODE REVIEW

File: include/dem/particle_particle_broad_search.h [hunk 1]
```diff
@@ -81,6 +78,24 @@ class ParticleParticleBroadSearch
...
+   *
+   * @param particle_begin The particle iterator to start storing particle id
+   * @param particles_to_evaluate The particle range to evaluate with the
+   * particle iterator prior storing ids
+   * @param contact_pair_candidates_container A vector which will contain all
+   * the particle pairs which are collision candidates
+   */
+  void
+  store_candidates(
+    typename Particles::ParticleHandler<dim>::particle_iterator_range::iterator
```
Comment (blaisb) [inline include/dem/particle_particle_broad_search.h:94]: I would pass a reference for the first one. Const if it is const.

File: include/dem/particle_particle_broad_search.h [hunk 2]
```diff
@@ -81,6 +78,24 @@ class ParticleParticleBroadSearch
       dim>::particle_particle_candidates &local_contact_pair_candidates,
     typename dem_data_containers::dem_data_structures<
       dim>::particle_particle_candidates &ghost_contact_pair_candidates);
+
+  /**
+   * Stores the candidate particle-particle collision pairs with a given
+   * particle iterator.
```
Comment (blaisb) [inline include/dem/particle_particle_broad_search.h:84]: You explain what the function does, but can you explain a bit what is is used for? Just give an example

File: source/dem/dem.cc [hunk 3]
```diff
@@ -963,6 +963,7 @@ DEMSolver<dim>::solve()
           // Reset checkpoint step
           checkpoint_step = false;
 
+          // Fill containers of particle-particle contact pair candidates
```
Comment (blaisb) [inline source/dem/dem.cc:966]:
```suggestion
          // Fill containers of particle-particle (local_contact_pair_candidates) and particle-ghost contact pair candidates (ghost_contact_pair_candidates) by using broad search
```

File: source/dem/find_cell_neighbors.cc [hunk 4]
```diff
@@ -39,8 +39,6 @@ FindCellNeighbors<dim>::find_cell_neighbors(
   // The reason is to find the cells located on the corners of the main cell.
   auto v_to_c = GridTools::vertex_to_cell_map(triangulation);
 
-  unsigned int cell_number_iterator = 0;
```
Comment (blaisb) [inline source/dem/find_cell_neighbors.cc:42]: Good catch.

File: source/dem/particle_particle_broad_search.cc [hunk 5]
```diff
@@ -19,11 +19,12 @@ ParticleParticleBroadSearch<dim>::find_particle_particle_contact_pairs(
   typename dem_data_containers::dem_data_structures<
     dim>::particle_particle_candidates &ghost_contact_pair_candidates)
 {
-  // First we will handle the local-lcoal candidate pairs
+  // First we will handle the local-local candidate pairs
```
Comment (blaisb) [inline source/dem/particle_particle_broad_search.cc:22]:
```suggestion
  // First we handle the local-local candidate pairs
```

File: source/dem/particle_particle_broad_search.cc [hunk 6]
```diff
@@ -19,11 +19,12 @@ ParticleParticleBroadSearch<dim>::find_particle_particle_contact_pairs(
...
   // Clearing local_contact_pair_candidates
   local_contact_pair_candidates.clear();
 
-  // Looping over cells_local_neighbor_list
+  // Looping over the list of cells with neighbors
```
Comment (blaisb) [inline source/dem/particle_particle_broad_search.cc:26]:
```suggestion
  // Looping over the potential cells which may contain particles. This includes the cell itself as well as the neighbouring cells that were identified.
```

File: source/dem/particle_particle_broad_search.cc [hunk 7]
```diff
@@ -156,41 +114,47 @@ ParticleParticleBroadSearch<dim>::find_particle_particle_contact_pairs(
...
                 }
             }
         }
     }
 }
 
+template <int dim>
+void
+ParticleParticleBroadSearch<dim>::store_candidates(
+  typename Particles::ParticleHandler<dim>::particle_iterator_range::iterator
```
Comment (blaisb) [inline source/dem/particle_particle_broad_search.cc:137]:
```suggestion
  const typename Particles::ParticleHandler<dim>::particle_iterator_range::iterator &
```

File: source/dem/particle_particle_broad_search.cc [hunk 8]
```diff
@@ -156,41 +114,47 @@ ParticleParticleBroadSearch<dim>::find_particle_particle_contact_pairs(
...
+    particle_begin,
+  typename Particles::ParticleHandler<dim>::particle_iterator_range
+    &                                 particles_to_evaluate,
```
Comment (blaisb) [inline source/dem/particle_particle_broad_search.cc:140]:
```suggestion
  const typename Particles::ParticleHandler<dim>::particle_iterator_range
    &                                 particles_to_evaluate,
```

File: source/dem/particle_particle_broad_search.cc [hunk 9]
```diff
@@ -156,41 +114,47 @@ ParticleParticleBroadSearch<dim>::find_particle_particle_contact_pairs(
...
+                                   particles_in_neighbor_cell,
+                                   ghost_contact_pair_candidates
+                                     [particle_in_main_cell->get_id()]);
                 }
             }
         }
     }
 }
 
+template <int dim>
```
Comment (blaisb) [inline source/dem/particle_particle_broad_search.cc:134]: This is a very very simple function. Can you move it to the .h and inline the function? This way you are 100% sure you won't loose performance ;)!

File: source/dem/particle_particle_broad_search.cc [hunk 10]
```diff
@@ -156,41 +114,47 @@ ParticleParticleBroadSearch<dim>::find_particle_particle_contact_pairs(
...
+  typename Particles::ParticleHandler<dim>::particle_iterator_range::iterator
+    particle_begin,
+  typename Particles::ParticleHandler<dim>::particle_iterator_range
+    &                                 particles_to_evaluate,
+  std::vector<types::particle_index> &contact_pair_candidates_container)
+{
+  // Create a temporary empty container
+  if (contact_pair_candidates_container.empty())
+    {
+      contact_pair_candidates_container.reserve(40);
```
Comment (blaisb) [inline source/dem/particle_particle_broad_search.cc:146]: 40 was a bit of an arbitrary number, it's good to see that now.... is it big is it small. We can play with it now that you did this refactoring

**Review by blaisb — CHANGES_REQUESTED** [review]
Very good work.  Very few changes. Good work

File: source/dem/particle_particle_broad_search.cc (hunk 10 above)
Comment (acdaigneault) [reply source/dem/particle_particle_broad_search.cc:146]: How? With the dp parameter? Or still arbitrary?

**Review by blaisb — APPROVED** [review]
Will merge once CI passes

**Pull request merged**